### ISSUE TITLE
Shell treats SyntaxError thrown by user code as an unfinished command

### ISSUE THREAD

**Issue by mattmccutchen** [meteor/meteor#8290]
When checking for a SyntaxError that indicates an unfinished command, the shell does not distinguish between SyntaxErrors from parsing the command and SyntaxErrors thrown by user code during the execution.  Example:

```
$ meteor create --bare .  # currently using METEOR@1.4.2.3
[...]
$ meteor npm install
[...]
$ meteor npm install --save babel-runtime
[...]
$ meteor
```

In another terminal:
```
$ meteor shell

Welcome to the server-side interactive shell!

Tab completion is enabled for global variables.

Type .reload to restart the server and the shell.
Type .exit to disconnect from the server and leave the shell.
Type .help for additional help.

> JSON.parse("[")
... 
```

~~Once this happens, typically the only way to recover is to interrupt and restart the shell.~~  This happened when I was debugging my application and was pretty confusing until I added a catch block to the shell command.  (Of course, the `JSON.parse("[")` was happening inside a lot of other code.)

**Comment by abernix** [comment]
Hmm, In attempting to reproduce, I'm dropped to the `...` (continuation) prompt in the same way as you demonstrate here, but when I press `Ctrl+C`, I'm returned to an active `>` prompt.   Am I missing something?

**Comment by mattmccutchen** [comment]
> when I press `Ctrl+C`, I'm returned to an active `>` prompt.

You're right.  I must not have tried this.  So there's an easy recovery.  The behavior is still confusing; I originally hit this with a longer command and was looking for the missing closing delimiter for a minute or two.

**Comment by abernix** [comment]
Yeah, the `node` REPL doesn't do this, but instead provides an more expected:

> SyntaxError: Unexpected end of input

It could be getting caught (`catch`) in a number of places.  Would need more investigation to figure it out specifically, but `shell-server` and `shell-client` code would be a place to start.

**Comment by Louis30** [comment]
Unsubscribe
On Jan 26, 2017 11:13 AM, "Jesse Rosenberger" <notifications@github.com>
wrote:

Yeah, the node REPL doesn't do this, but instead provides an more expected:

SyntaxError: Unexpected end of input

It could be getting caught (catch) in a number of places. Would need more
investigation to figure it out specifically, but shell-server and
shell-client code would be a place to start.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/meteor/meteor/issues/8290#issuecomment-275482800>, or mute
the thread

**Comment by benjamn** [comment]
Can this be fixed with an update to the `shell-server` package, or do we need to update the client (i.e. `meteor-tool`)?

**Comment by abernix** [comment]
I had reason to believe it was a client-side fix, but I hadn't got the chance to investigate/debug.

@mattmccutchen If you get a chance to look into this, please let us know!  Happy to help in the debugging process if you need assistance!

**Comment by mattmccutchen** [comment]
This turned out to be an easy server-side fix: #8446.

**Comment by abernix** [comment]
This should be fixed in Meteor 1.4.3.2.  You can try the latest 1.4.3.2 beta and help confirm by running:

    meteor update --release 1.4.3.2-beta.0

Please report back if you encounter any problems, and thanks for reporting **and** taking care of this!
